### PR TITLE
adds support for actionId

### PR DESCRIPTION
**Pull request by vinistevam**
### `@metamask/package-a`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

### `@metamask/package-b`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/transaction-controller/src/TransactionController.ts [hunk 1]
```diff
@@ -377,10 +396,14 @@ export class TransactionController extends BaseController<
...
+    if (!existingTransactionMeta) {
+      transactions.push(transactionMeta);
+      this.update({
+        transactions: this.trimTransactionsForState(transactions),
+      });
+      this.hub.emit(`unapprovedTransaction`, transactionMeta);
+    }
 
     return {
       result: this.processApproval(transactionMeta),
```
Comment (OGPoyraz) [inline packages/transaction-controller/src/TransactionController.ts:409]: A bit confused here, shouldn't we somehow skip creating another approval if it's an `existingTransaction`?

File: packages/transaction-controller/src/TransactionController.ts [hunk 2]
```diff
@@ -1049,18 +1089,29 @@ export class TransactionController extends BaseController<
...
+   * @param opts - Additional options to control how the transaction is added.
+   * @param opts.actionId - Unique ID to prevent duplicate requests
    */
-  private cancelTransaction(transactionID: string) {
-    const transactionMeta = this.state.transactions.find(
-      ({ id }) => id === transactionID,
-    );
+  private cancelTransaction(
+    transactionID: string,
+    opts: { actionId?: string } = {},
```
Comment (OGPoyraz) [inline packages/transaction-controller/src/TransactionController.ts:1097]: Nit: we could use the same parameters convention with `approveTransaction`

File: packages/transaction-controller/src/types.ts [hunk 3]
```diff
@@ -43,6 +44,7 @@ type TransactionMetaBase = {
   blockNumber?: string;
   deviceConfirmedOn?: WalletDevice;
   verifiedOnBlockchain?: boolean;
+  actionId?: string;
```
[the remark this answers is not in the document]
Comment (vinistevam) [reply packages/transaction-controller/src/types.ts:47]: That makes sense, I'm pushing one change based on your comment so I will also sort these properties :100:

File: packages/transaction-controller/src/TransactionController.ts [hunk 4]
```diff
@@ -331,23 +331,38 @@ export class TransactionController extends BaseController<
     }
   }
 
+  /**
+   * Get transaction with provided actionId.
+   *
+   * @param actionId - Unique ID to prevent duplicate requests
+   * @returns the filtered transaction
+   */
+  getTransactionWithActionId(actionId?: string) {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:340]: Can this be private in order to minimise the public API as much as possible?

File: packages/transaction-controller/src/TransactionController.ts [hunk 5]
```diff
@@ -1049,18 +1085,24 @@ export class TransactionController extends BaseController<
...
    * @param transactionID - The ID of the transaction to cancel.
+   * @param actionId - Unique ID to prevent duplicate requests
    */
-  private cancelTransaction(transactionID: string) {
-    const transactionMeta = this.state.transactions.find(
-      ({ id }) => id === transactionID,
-    );
+  private cancelTransaction(transactionID: string, actionId?: string) {
+    const transactionMeta = this.state.transactions.find((transaction) => {
+      if (actionId) {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:1092]: Is this check required? Won't the transaction ID still be the same if it's a duplicate call as we've got the same transaction meta thus the same ID when the code reaches here?

File: packages/transaction-controller/src/TransactionController.ts [hunk 6]
```diff
@@ -961,7 +995,9 @@ export class TransactionController extends BaseController<
     const { transactions } = this.state;
     const releaseLock = await this.mutex.acquire();
     const { chainId } = this.getChainAndNetworkId();
-    const index = transactions.findIndex(({ id }) => transactionID === id);
+    const index = transactions.findIndex((transaction) => {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:998]: Is this needed or should we dereference when possible?

File: packages/transaction-controller/src/TransactionController.ts [hunk 7]
```diff
@@ -890,6 +923,7 @@ export class TransactionController extends BaseController<
   ): Promise<string> {
     const transactionId = transactionMeta.id;
     let resultCallbacks: AcceptResultCallbacks | undefined;
+    const { actionId } = transactionMeta;
 
     try {
       const acceptResult = await this.requestApproval(transactionMeta, {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:929]: One key addition I think we need is to skip additional approvals if we are a duplicate call.

The extension does this by passing an `isExisting` argument to this method, skipping the approval if true, and then waiting for a suitable event before resolving the promise.

Unless of course the status of the meta is already final in which case we resolve immediately.

File: packages/transaction-controller/src/TransactionController.ts [hunk 8]
```diff
@@ -1255,11 +1297,11 @@ export class TransactionController extends BaseController<
     return String(txMeta.id);
   }
 
-  private isTransactionCompleted(transactionid: string): {
+  private isTransactionCompleted(transactionID: string): {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:1300]: For naming going forward, I think the standard is `transactionId` in code, and `Transaction ID` in labels.

File: packages/transaction-controller/src/TransactionController.ts (hunk 5 above)
Comment (matthewwalsh0) [reply packages/transaction-controller/src/TransactionController.ts:1092]: I'm wondering if you meant to make this change to `stopTransaction` which confusingly is equivalent to `createCancelTransaction` in the extension?

In core at least, `cancelTransaction` is reject an unapproved transaction, and `stopTransaction` is cancel an already submitted transaction.

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 9]
```diff
@@ -686,6 +687,53 @@ describe('TransactionController', () => {
       );
     });
 
+    it.each([
+      [
+        'adds single unapproved transactions with same actionId',
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:692]:
```suggestion
        'does not add duplicate transaction if actionId already used',
```

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 10]
```diff
@@ -686,6 +687,53 @@ describe('TransactionController', () => {
...
+        ACTION_ID_MOCK,
+        ACTION_ID_MOCK,
+        1,
+      ],
+      [
+        'adds multiple unapproved transactions with different actionId',
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:698]:
```suggestion
        'adds additional transaction if actionId not used',
```

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 11]
```diff
@@ -686,6 +687,53 @@ describe('TransactionController', () => {
...
+            from: ACCOUNT_MOCK,
+            to: ACCOUNT_MOCK,
+          },
+          {
+            origin: mockOrigin,
+            actionId: firstActionId,
+          },
+        );
+
+        await controller.addTransaction(
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:721]: Following my action about skipping approvals and waiting for the event, we'll need another test to check that the duplicate transaction result doesn't resolve until the first transaction result resolves.

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 12]
```diff
@@ -686,6 +687,53 @@ describe('TransactionController', () => {
       );
     });
 
+    it.each([
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:690]: To keep this tests tidy, should we add a describe for these such as `with actionId`?

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 9 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts]: I moved the tests to their own describe, so once I tried to use the commit suggestion it does not work, but I've included this improvement manually.

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 10 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts]: Same as previous, included manually.

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 12 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts:701]: Perfect, moved all the tests to a separated describe.

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 11 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts:732]: Included this additional test to ensure the duplicate transaction result does not resolve until the first transaction.

File: packages/transaction-controller/src/TransactionController.ts (hunk 7 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.ts]: Included `isExisting` argument in `processApproval` and skipping the approval creation if `true` as in the extension.

File: packages/transaction-controller/src/TransactionController.ts (hunk 5 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.ts]: I was reviewing my changes in the `cancelTransaction` after reading your clarification and the equivalence between the extension got me confused and misled me with this additional change. I reverted this part.

**Review by OGPoyraz — APPROVED** [review]

File: packages/transaction-controller/src/TransactionController.ts [hunk 13]
```diff
@@ -570,12 +582,21 @@ export class TransactionController extends BaseController<
...
+   * @param opts.actionId - Unique ID to prevent duplicate requests
    */
   async speedUpTransaction(
     transactionID: string,
     gasValues?: GasPriceValue | FeeMarketEIP1559Values,
+    opts: { actionId?: string } = {},
   ) {
+    const { actionId } = opts;
+    // If transaction is found for same action id, do not create a new speed up transaction.
+    if (this.getTransactionWithActionId(actionId)) {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:596]: Do we not need identical logic on `stopTransaction` too?

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 14]
```diff
@@ -698,6 +699,138 @@ describe('TransactionController', () => {
...
+        })
+        .catch(() => undefined);
+
+      await firstResult;
+      const { transactions } = controller.state;
+
+      expect(transactions).toHaveLength(1);
+      expect(messengerMock.call).toHaveBeenCalledTimes(1);
+      expect(firstResolved).toBe(true);
+      expect(secondResolved).toBe(false);
```
[the remark this answers is not in the document]
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts]: Thanks for the reference test in the extension. I've changed the test based on it approving now via `delayMessegerMock`.

File: packages/transaction-controller/src/TransactionController.ts (hunk 13 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.ts:607]: I was confused by adding this logic to `stopTransaction`maily because it does not include the `rawTransaction` into the state so the second time calling it it will create the transaction anyway because in the first call the `actionId` was not saved. I will ping you so we can double-check this point together.

File: packages/transaction-controller/src/TransactionController.ts (hunk 7 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.ts]: Great! Now it's waiting for `${transactionMeta.id}:finished` event as in the extension. It also checks if the transaction is not finished to skip sign and submit and instead hold the promised status is finished.

File: packages/transaction-controller/src/TransactionController.ts [hunk 15]
```diff
@@ -1426,6 +1457,28 @@ export class TransactionController extends BaseController<
...
+   * @param actionId - Unique ID to prevent duplicate requests
+   * @returns the filtered transaction
+   */
+  private getTransactionWithActionId(actionId?: string) {
+    return this.state.transactions.find(
+      (transaction) => actionId && transaction.actionId === actionId,
+    );
+  }
+
+  async _waitForTransactionFinished(
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:1473]: As we're in core, we can omit the underscore and use the `private` keyword.

File: packages/transaction-controller/src/TransactionController.ts [hunk 16]
```diff
@@ -901,46 +923,55 @@ export class TransactionController extends BaseController<
...
-      if (meta && !isCompleted) {
-        await this.approveTransaction(transactionId);
-      }
-    } catch (error: any) {
-      const { meta, isCompleted } = this.isTransactionCompleted(transactionId);
-
-      if (meta && !isCompleted) {
-        if (error.code === errorCodes.provider.userRejectedRequest) {
-          this.cancelTransaction(transactionId);
+        const { isCompleted: isTxCompleted } =
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:951]: As we're now checking for completion above, is this additional check needed?

If we know it wasn't completed before requesting approval, it is safe to assume that's still the case once approved?

File: packages/transaction-controller/src/TransactionController.ts [hunk 17]
```diff
@@ -901,46 +923,55 @@ export class TransactionController extends BaseController<
...
-          throw ethErrors.provider.userRejectedRequest(
-            'User rejected the transaction',
-          );
-        } else {
-          this.failTransaction(meta, error);
+        if (!isTxCompleted) {
+          await this.approveTransaction(transactionId);
+        }
+      } catch (error: any) {
+        const { isCompleted: isTxCompleted } =
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:958]: I appreciate we always have a reference to `meta` now from the start of the method, but just to be sure, in case some error or code path has removed our original metadata, should we keep the local reference here and check that instead?

File: packages/transaction-controller/src/TransactionController.ts [hunk 18]
```diff
@@ -901,46 +923,55 @@ export class TransactionController extends BaseController<
...
+          this.isTransactionCompleted(transactionId);
+        if (meta && !isTxCompleted) {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:960]: If a `meta` reference is guaranteed now, should we remove this `meta` condition?

File: packages/transaction-controller/src/types.ts [hunk 19]
```diff
@@ -13,6 +13,11 @@ export type TransactionMeta =
  * Information about a single transaction such as status and block number.
  */
 type TransactionMetaBase = {
+  /**
+   * Unique ID to prevent duplicate requests
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/types.ts:17]:
```suggestion
   * Unique ID to prevent duplicate requests.
```

File: packages/transaction-controller/src/TransactionController.ts [hunk 20]
```diff
@@ -382,6 +382,7 @@ export class TransactionController extends BaseController<
    *
    * @param transaction - The transaction object to add.
    * @param opts - Additional options to control how the transaction is added.
+   * @param opts.actionId - Unique ID to prevent duplicate requests
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:385]:
```suggestion
   * @param opts.actionId - Unique ID to prevent duplicate requests.
```

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 21]
```diff
@@ -802,6 +967,27 @@ describe('TransactionController', () => {
       });
     });
 
+    it('fails to estimate gas', async () => {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:970]: Why are we adding unrelated unit tests? Is it to stop the coverage decreasing?

If so, would it be simpler to let the coverage go down slightly in order to minimise the size and clarity of the PR, and then we can add additional unit tests in a separate dedicated PR if needed?

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 22]
```diff
@@ -704,6 +705,170 @@ describe('TransactionController', () => {
     });
   });
 
+  describe('with actionId', () => {
+    it('adds multiple transactions with same actionId and ensures second transaction result does not resolves before the first transaction result', async () => {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:709]: Can we split this into two tests as in the extension?

The first ignores the `result` promise and just checks the number of transactions in the state, where as the second would only assert the results are resolved together?

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 23]
```diff
@@ -704,6 +705,170 @@ describe('TransactionController', () => {
...
+          actionId: ACTION_ID_MOCK,
+        },
+      );
+      secondResult
+        .then(() => {
+          secondTransactionCompleted = true;
+        })
+        .catch(() => undefined);
+
+      expect(firstTransactionCompleted).toBe(false);
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:749]: Just to be thorough, can we add a `flushPromises` before these assertions to ensure the event loop has had chance to run our `then` callbacks if the results had already resolved?

File: packages/transaction-controller/src/TransactionController.ts (hunk 18 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.ts]: It still can be `undefined` so to avoid `failTransaction` I can move the check to the first `if`.

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 21 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts]: Make sense, I will remove it and I will create a separate PR to increase the coverage.

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 22 above)
Comment (vinistevam) [reply packages/transaction-controller/src/TransactionController.test.ts:754]: Perfect! I've split the tests.

**Review by matthewwalsh0 — APPROVED** [review]

**Comment by vinistevam** [comment]
> @vinistevam The linked issue [MetaMask/MetaMask-planning#1055](https://github.com/MetaMask/MetaMask-planning/issues/1055) is private an non-readable for general contributors. Could it be moved/copied to a public issue in this repo?

I was talking with the team about it and to avoid confusion I'm removing the ticket reference from the description and including more context in the PR.

**Review by OGPoyraz — APPROVED** [review]
